pkg/registrymirror/registrymirrorsstore.go

```diff
@@ -0,0 +1,325 @@
+	DeleteImageTagMirrorSet(uid types.UID) error
+
+	// PullSources will return image references in the order they should be attempted when pulling the image.
+	// The returned slice may (or may not) include the source image based on the registries config.
+	// Source image must be fully qualified (include the registry hostname).
+	PullSources(srcImage string) ([]string, error)
+}
+
+// FileStore stores/reads the consolidated registries config from a filesystem.
+type FileStore struct {
+	configPath string
+
+	// holds the mirror sets used for managing registry mirrors.
+	icspRules   map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy
+	idmsRules   map[types.UID]*configV1.ImageDigestMirrorSet
+	itmsRules   map[types.UID]*configV1.ImageTagMirrorSet
+	ruleRWMutex sync.RWMutex
+
+	// timer is used when performing delayed writes to filesystem.
+	timer       *time.Timer
+	timerMutex  sync.Mutex
+	updateDelay time.Duration
+
+	systemContext *ciTypes.SystemContext
+}
+
+var _ Store = (*FileStore)(nil)
+
+// fileStoreOption is used to provide functional options to the fileStore.
+type fileStoreOption func(*FileStore)
+
+// WithConfigPath sets the path to read/write the consolidated registries config.
+func WithConfigPath(path string) fileStoreOption {
+	return func(s *FileStore) { s.configPath = path }
+}
+
+// WithDelay sets a delay before the updated config is written to disk.
+func WithDelay(delay time.Duration) fileStoreOption {
+	return func(s *FileStore) { s.updateDelay = delay }
+}
+
+// NewFileStore creates a new FileStore.
+func NewFileStore(opts ...fileStoreOption) *FileStore {
+	s := &FileStore{
+		configPath:  defaultRegistriesPath,
+		updateDelay: defaultDelay,
+		icspRules:   make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy),
+		idmsRules:   make(map[types.UID]*configV1.ImageDigestMirrorSet),
+		itmsRules:   make(map[types.UID]*configV1.ImageTagMirrorSet),
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	s.systemContext = &ciTypes.SystemContext{SystemRegistriesConfPath: s.configPath}
+	return s
+}
+
+// Cleanup resets the store which includes in-memory and disk resources.
+func (s *FileStore) Cleanup() {
+	s.ruleRWMutex.Lock()
+	defer s.ruleRWMutex.Unlock()
+
+	s.icspRules = make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy)
+	s.idmsRules = make(map[types.UID]*configV1.ImageDigestMirrorSet)
+	s.itmsRules = make(map[types.UID]*configV1.ImageTagMirrorSet)
+
+	if err := os.Remove(s.configPath); err != nil && !os.IsNotExist(err) {
+		log.Warnf("Failed to cleanup registries config at %q: %v", s.configPath, err)
+	}
+	sysregistriesv2.InvalidateCache()
+}
+
+func (s *FileStore) updateConfig() error {
+	if s.updateDelay == 0 {
+		return s.updateConfigNow()
+	}
+
+	return s.updateConfigDelayed()
+}
+
+func (s *FileStore) updateConfigDelayed() error {
+	// Using a different mutex to protect timer vs. rules to avoid deadlock.
+	s.timerMutex.Lock()
+	defer s.timerMutex.Unlock()
+
+	if s.timer != nil {
+		s.timer.Stop()
```

Comment:
Should this be called in `Cleanup` as well?

Reply:
Shouldn't need to, because am clearing out the rules and then erasing the config file while holding the rules lock, even if the timer fires after the fact it will write an empty file because the rules have been cleared - so no harm done.

If we were OK with holding multiple locks at the same time I would add it to Cleanup - but because we prefer not to do that cannot guarantee will be in sync (something may always squeeze in the middle).

Something else to consider is that when these CRs are deleted, or modified - they often trigger the k8s nodes to drain all pods. So even if this gets out of sync - the node drain will cause sensor to restart and be back to a clean state.

Reply:
I removed the timer and replaced with the ACS Signal/AfterFunc per @dhaus67 's suggestion. As part of that added canceling any pending update to the cleanup function.